### ISSUE TITLE
[Proposal] JSON::Any all the way down

### ISSUE THREAD

**Issue by feifanzhou** [crystal-lang/crystal#3158]
Currently, trying to work with a JSON collection wrapped as a `JSON::Any` will give me the raw `JSON::Type` one level down:

``` cr
JSON.parse("[{\"key\": \"value\"}]").as_a
```

This returns `Array(JSON::Type)`. If I wanted to work with that inner hash, I'd have to manually wrap it into another `JSON::Any`:

``` cr
JSON.parse("[{\"key\": \"value\"}]").as_a.map{ |h| JSON::Any.new(h) }
```

This will give me a `Hash(String, JSON::Type)`. Repeat for working with values in the hash. 

Instead, can we have `as_a` return `Array(JSON::Any)` and `as_h` return `Hash(String, JSON::Any)` so they behave more intuitively? The JSON types will "disappear" a bit more, leading to more intuitive behavior that more closely matches `JSON.parse` behavior in Ruby.

If there actually _is_ an easy way to chain or automatically convert `JSON::Type` that I've missed, that's awesome! Maybe we could make it more obvious in the docs :)

**Comment by jhass** [comment]
I agree, it should either be recursive or we should get rid of it entirely.

**Comment by asterite** [comment]
I'm thinking more and more of removing it, as it makes it difficult to understand what's going on. Dealing with unions is a bit more cumbersome, but it's something you do all the time.

The problem with returning `Array(JSON::Any)` is that we'll have to create a separate array, using `map`, to wrap the values, so the performance will be horrible.

**Comment by jhass** [comment]
Well, one *could* (I'm not saying should) workaround that by creating `JSON::Any::Array` and `JSON::Any::Hash` which wrap the array/hash instead.

**Comment by feifanzhou** [comment]
@asterite if `JSON::Any` is removed, would it be possible to add the `as_*` methods to `JSON::Type` itself? Otherwise, seems like it would require the dev to write a lot of boilerplate type checking and casting when working with JSON

**Comment by asterite** [comment]
```cr
JSON.parse("...").as(Array)[0].as(Hash)["foo"]
```

I think before making the syntax change to `as` this was a lot more tedious:

```cr
((JSON.parse("...") as Array)[0] as Hash)["foo"]
```

When I was editing that I had to go to the beginning of the line twice to add the matching parentheses. With the new syntax it's more fluent, so probably less cumbersome (the first snippet felt natural to write, the second didn't)

**Comment by feifanzhou** [comment]
Ah, didn't know we had that! Combined with some type aliases that looks just as good as the `as_*` that `JSON::Any` has now.

**Comment by ysbaddaden** [comment]
I tried several times to use `JSON::Any` and `YAML::Any`; except for very basic cases like an array of strings, they always get into the way, and the `as_` methods are destabilizing. Simple casting on the raw object or a pull parser always led to a better design.

I vote to remove them, or at least introduce `parse_raw` methods.

**Comment by BlaXpirit** [comment]
Regarding the original post, that just looks like misunderstanding. It really is `JSON::Any` all the way down right now, as long as you don't call the `.to_*` methods until the very end.

**Comment by jhass** [comment]
@BlaXpirit No, it's not https://carc.in/#/r/176d

**Comment by BlaXpirit** [comment]
@jhass, well, you used the `.to_a` method, then of course immediately it's not.

**Comment by BlaXpirit** [comment]
https://carc.in/#/r/176r

**Comment by RX14** [comment]
The point to me of `JSON::Any` is for hacking together a quick prototype which needs JSON parsing, where you really don't want to bother mapping your JSON to objects, you just want the data and you don't really need error handling, or speed.`JSON::Any` fulfils that need to hack together a quick API call to make it work, and so I think it should stay.

**Comment by ozra** [comment]
You don't need to map to objects. See asterites example above. Doing the casts before accessing a value is simple, and a good basic verification you're using the API right - it's not much different from using the `to_a` etc. style.

**Comment by RX14** [comment]
@ozra `JSON::Any` is effectively a wrapper around the casts which lets you perform basic operations on the value by asserting if the operator you are using works on the contained type. Using `.as` and using `JSON::Any` have the same levels of safety because they both assert the same things, except one is implicit and easier to use and read.

**Comment by asterite** [comment]
Fixed by #5183
